nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java

```diff
@@ -718,6 +714,220 @@ public void varargsParameter() {
+            "   @Nullable R apply(P p);",
+            "  }",
+            " static class TestFunc implements Fn<String, String> {",
+            "  @Override",
+            "  //This override is fine and is handled by the current code",
+            "  public @Nullable String apply(String s) {",
+            "   return s;",
+            "  }",
+            " }",
+            " static void useTestFunc(String s) {",
+            "  Fn<String, String> f = new TestFunc();",
+            "  String t = f.apply(s);",
+            "  // BUG: Diagnostic contains: dereferenced expression",
+            "  t.hashCode();",
+            " }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void overrideWithNestedTypeParametersInReturnType() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.jspecify.annotations.Nullable;",
+            "class Test {",
+            " class P<T1 extends @Nullable Object, T2 extends @Nullable Object>{}",
+            " interface Fn<T extends P<R, R>, R extends @Nullable Object> {",
+            "  T apply();",
+            " }",
+            " class TestFunc1 implements Fn<P<@Nullable String, String>, @Nullable String> {",
+            " @Override",
+            "  // BUG: Diagnostic contains: Method returns P<@Nullable String, @Nullable String>, but overridden method",
+            " public P<@Nullable String, @Nullable String> apply() {",
+            "   return new P<@Nullable String, @Nullable String>();",
+            "  }",
+            " }",
+            " class TestFunc2 implements Fn<P<@Nullable String, @Nullable String>, @Nullable String> {",
+            " @Override",
+            "  // BUG: Diagnostic contains: Method returns P<@Nullable String, String>, but overridden method returns",
+            " public P<@Nullable String, String> apply() {",
+            "   return new P<@Nullable String, String>();",
+            "  }",
+            " }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void overrideWithNestedTypeParametersInParameterType() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.jspecify.annotations.Nullable;",
+            "class Test {",
+            "  class P<T1 extends @Nullable Object, T2 extends @Nullable Object>{}",
+            " interface Fn<T extends P<R, R>, R extends @Nullable Object> {",
+            "  String apply(T t, String s);",
+            " }",
+            " class TestFunc implements Fn<P<String, String>, String> {",
+            " @Override",
+            "  // BUG: Diagnostic contains: Parameter has type P<@Nullable String, String>, but overridden method has parameter type P<String, String>",
+            "  public String apply(P<@Nullable String, String> p, String s) {",
+            "    return s;",
+            "  }",
+            " }",
+            "}")
+        .doTest();
+  }
```

Comment:
I added a test in 8bca16e16692067b11eb27907ba791cb36e2a9d0.  The good news is that we get an understandable behavior here; the `@Contract` annotations apply only to the method directly being annotated.  The not-so-great news is that we don't really have a story in terms of `@Contract` and inheritance.  So, for the case where `@Contract` is written on `Fn.apply`, and `TestFunc.apply` does _not_ adhere to the contract, we will get no error message, even with contract checking enabled.  Maybe we should file a follow-up issue on that?  In any case, that issue is independent of generics, though it may get more complex to do the "right" checking when accounting for generic types

Reply:
Did we end up tracking this in an issue somewhere?

Reply:
Created #803 for this